openflow-sdn/openflow-protocol.md

```diff
@@ -0,0 +1,65 @@
+
+## Basics of Openflow
+
+As mentioned, the OpenFlow protocol lays out the foundation for communication between an SDN controller and a dumb network device. This protocol was developed first by researchers at Stanford University in 2008 and was first adopted by [Google](https://static.googleusercontent.com/media/research.google.com/en//pubs/archive/42948.pdf) in their backbone network in 2011-2012. It is managed now by the [Open Networking Foundation(ONF)](https://opennetworking.org/sdn-definition/). The latest version used in the industry is V1.5. 
+
+OpenFlow is the standard southbound protocol used between the SDN controller and the switch. The SDN controller takes the information from the applications and converts them into flow entries which are fed to the switch via OF. It can also be used for monitoring switch and port statistics for network management.
+
+**NOTE**
+The OpenFlow protocol is only established between a controller and the switch. It does not affect the rest of the network and if a packet capture were to be taken between two switches in a network, both of them connected to the controller via another port, the packet capture would not reveal any OF messages between the switches. It is strictly for the use between a switch and the controller. The rest of the network is not affected.
+
+### Initiation of OpenFlow channel 
+
+OpenFlow protocol works on the TCP protocol. The standard protocol is TCP 6633 for OF V1.0 and 6653 for OF V1.3+. There needs to be IP connectivity between the controller and the switches for them to establish an OF connection. OF channel is formed only after a successful TCP 3-way handshake. 
+
+1. To start the OF channel communication, the switch sends a "HELLO" packet to introduce it to the controller. The switch also sends information like the highest version of OF it supports. The controller replies to the hello message with its highest supported OF version. Then, the switch negotiates on the highest level of the OpenFlow version that they both support. 
+2. Once the version is negotiated, the controller sends a "FEATURE_REQUEST" message. This message essentially asks the switch for its supported OF capabilities like the number of flow tables supported, supported actions, etc. The switch replies to it with a "FEATURE_REPLY" message stating all its capabilities along with its unique identifier or Datapath ID(DPID).
+
+After this, it is said that the OpenFlow channel is successfully established between the switch and the controller. 
+
+The connection between the controller and switch is really important as it is the only way for a switch to communicate with a controller. To secure this connection, a protocol like TLS can also be used instead of a TCP connection. Here, the controller and switch need to have the proper certificates and keys for a successful TLS connection. This prevents snooping on the OF channel.
```

Comment:
```suggestion
The connection between the controller and switch is essential as it is the only way for a switch to communicate with a controller. To secure this connection, a protocol like TLS can also be used instead of a TCP connection. Here, the controller and switch need to have the proper certificates and keys for a successful TLS connection. This prevents snooping on the OF channel.
```